### ISSUE TITLE
Error in Test Scripts

### ISSUE THREAD

**Issue by hkarakurt8742** [BIMIB-DISCo/scFBA#2]
Hello,
I run the scripts scFBA_test.m and replicateScFBAresults.m but I have an error in single2IntPopModel step.
Error I have:
Error using parseCobraVarargin (line 129)
The value of 'osenseStr' is invalid. It must satisfy the function: @(x)strcmp(x,'max')|strcmp(x,'min').

Error in fluxVariability (line 115)
[funParams, cobraParams, solverVarargin] = parseCobraVarargin(varargin, optArgin, defaultValues, validator, problemTypes,
'solverParams', true);

Error in integrateRAS (line 32)
[Fmin,FMax] = fluxVariability(modelPop, 100, '', RxnList); %controllare che 100 vada bene

Error in single2IntPopModel (line 87)
    scStruct = integrateRAS(scStruct, 'modelDel');

Error in scFBA_test (line 51)
BC04 = single2IntPopModel(BC04, HMRcore, IdxExRxns, IdxCoopRxn , 's');

Error in run (line 91)
evalin('caller', strcat(script, ';'));

Is scFBA only works with Cobra Toolbox 2017 or later than 2017 is okay?

Thank you in advance

**Comment by hkarakurt8742** [comment]
Cobra v 2.0 solved that problem but now I have this one

Error using extractGeneRow (line 52)
Gene ID dictionary is not compleate, please check

Error in scFBA_test (line 22)
BC04_filt = extractGeneRow(HMRcore, BC04dataset, 1, DictCORE);

I am not sure why I have this error. Everything looks okay in data.

**Comment by DavideMaspero** [comment]
Dear @hkarakurt8742,

That error is raised when the dictionary does not contain an entry for each gene in model.genes. Unfortunately, I was not able to replicate your issue as I have just cloned the repository and run the scFBA_test script and I did not have any problem. 

Could you try to clear your workspace and run again scFBA_test?

**Comment by hkarakurt8742** [comment]
Thank you for your answer,
Actually I found the problem. When I directly read hmr_core.xml model with readCbModel function I have this error because genes in model and dictionary are different.
I loaded .mat file of the model and I replicated publication results.

As I understand from the publication, cooperation reactions are based on intra-tumor cooperation phenomena. 
HMR Core model have cooperation reactions but original HMR and HMR2 does not have reaction names. Also other models such as Recon2 do not have. 
Can reactions of metabolite transfer to extracellular are considered as COOP reactions?

Thank you for your kind answer and help again.

**Comment by DavideMaspero** [comment]
Yes, you can specify as a cooperation reaction any exchange reaction present in your model.
For example, consider the following exchange reaction
metabolite_in[c] <=> ""

In the result population model (considering 2 subpopulations) three reactions will be added:
metabolite_in_1[c] <=> metabolite_out[s]
metabolite_in_2[c] <=> metabolite_out[s]
metabolite_out[s] <=> ""
where 's' refers to the compartment shared among the subpopulations. 

Hope this will help you.

**Comment by SantiagoMille** [comment]
> Cobra v 2.0 solved that problem but now I have this one
> 
> Error using extractGeneRow (line 52)
> Gene ID dictionary is not compleate, please check
> 
> Error in scFBA_test (line 22)
> BC04_filt = extractGeneRow(HMRcore, BC04dataset, 1, DictCORE);
> 
> I am not sure why I have this error. Everything looks okay in data.

@hkarakurt8742 did you downgrade to Cobra v2.0? I'm currently using the latest version ( v3.0 ) and I am having the same problem. Also, just wondering how long does it take for you to run the whole script? 

Best,

**Comment by hkarakurt** [comment]
> > Cobra v 2.0 solved that problem but now I have this one
> > Error using extractGeneRow (line 52)
> > Gene ID dictionary is not compleate, please check
> > Error in scFBA_test (line 22)
> > BC04_filt = extractGeneRow(HMRcore, BC04dataset, 1, DictCORE);
> > I am not sure why I have this error. Everything looks okay in data.
> 
> @hkarakurt8742 did you downgrade to Cobra v2.0? I'm currently using the latest version ( v3.0 ) and I am having the same problem. Also, just wondering how long does it take for you to run the whole script?
> 
> Best,

Yes, I did exactly what you said. Everything looks fine with Cobra v2.0. As I remember it took about 2 hours.

Best,

**Comment by SantiagoMille** [comment]
> > > Cobra v 2.0 solved that problem but now I have this one
> > > Error using extractGeneRow (line 52)
> > > Gene ID dictionary is not compleate, please check
> > > Error in scFBA_test (line 22)
> > > BC04_filt = extractGeneRow(HMRcore, BC04dataset, 1, DictCORE);
> > > I am not sure why I have this error. Everything looks okay in data.
> > 
> > 
> > @hkarakurt8742 did you downgrade to Cobra v2.0? I'm currently using the latest version ( v3.0 ) and I am having the same problem. Also, just wondering how long does it take for you to run the whole script?
> > Best,
> 
> Yes, I did exactly what you said. Everything looks fine with Cobra v2.0. As I remember it took about 2 hours.
> 
> Best,

Thanks a lot.

**Comment by CJQ-CS-WHU** [comment]
Hello, I'd like to ask yourwhere I can download the Codra v2.0. 
I look for it the 'Add-On Explorer' in matlab but there is no history version for downloading.   
I will be appreaciate if you can tell me where I can get it! Thanks.

**Comment by CJQ-CS-WHU** [comment]
> > > Cobra v 2.0解决了该问题，但现在我
> > > 使用extractGeneRow（第52行）遇到了这个错误。
> > > 基因ID字典不完整，请检查
> > > scFBA_test中的错误（第22行）
> > > BC04_filt = extractGeneRow（HMRcore，BC04dataset，1，DictCORE）; 
> > > 我不确定为什么会有这个错误。数据看起来一切正常。
> > 
> > 
> > @ hkarakurt8742您是否已降级到Cobra v2.0？我当前正在使用最新版本（v3.0），并且遇到了同样的问题。另外，只是想知道运行整个脚本需要多长时间？
> > 最好，
> 
> 是的，我完全按照你说的做。眼镜蛇v2.0一切看起来都很好。我记得花了大约2个小时。
> 
> 最好，



> > > > Cobra v 2.0 solved that problem but now I have this one
> > > > Error using extractGeneRow (line 52)
> > > > Gene ID dictionary is not compleate, please check
> > > > Error in scFBA_test (line 22)
> > > > BC04_filt = extractGeneRow(HMRcore, BC04dataset, 1, DictCORE);
> > > > I am not sure why I have this error. Everything looks okay in data.
> > > 
> > > 
> > > @hkarakurt8742 did you downgrade to Cobra v2.0? I'm currently using the latest version ( v3.0 ) and I am having the same problem. Also, just wondering how long does it take for you to run the whole script?
> > > Best,
> > 
> > 
> > Yes, I did exactly what you said. Everything looks fine with Cobra v2.0. As I remember it took about 2 hours.
> > Best,
> 
> Thanks a lot.

Hello, I got invovled in a problem that I don't know how to get  Cobra v2.0. Can you give me some hint? Thank you very much!

**Comment by riddhimandhar** [comment]
Hello, 
When we run the scFBA_test using cobra toolbox v2.0, we get the following error - 

Unrecognized function or variable 'buildRxnGeneMat'.

Error in ScFBAExpSetting (line 122)
model = buildRxnGeneMat(model); %generate rxnGeneMat field

Error in scFBA_test (line 47)
HMRcore = ScFBAExpSetting(HMRcore, length(BC04.CellType));

I am using the cobra toolbox from here and it seems that the buildRxnGeneMat function is not there in this version https://github.com/opencobra/cobratoolbox/releases/tag/v2.0.0. 

I could find this function in cobra toolbox v3.0. 

Could you please specify the exact version of cobra toolbox that we should use? 

Thanks a lot.

**Comment by PratikPathade** [comment]
I am trying to run this command 
HMRcore = ScFBAExpSetting(HMRcore, length(BC04.CellType));
got an error like this:- 
Error using EditBoundaries
Too many input arguments.

Error in ScFBAExpSetting (line 30)
model = EditBoundaries(model, 'SLC25A5_SLC25A4_SLC25A6', 0,10000, false, false);

I am not sure why I am getting this error.
and in the next step.
BC04 = single2IntPopModel(BC04, HMRcore, IdxExRxns, IdxCoopRxn, 's');

Error:- 
Error using parseCobraVarargin
The value of 'osenseStr' is invalid. It must satisfy the function: @(x)strcmp(x,'max')|strcmp(x,'min').

Error in fluxVariability (line 115)
[funParams, cobraParams, solverVarargin] = parseCobraVarargin(varargin, optArgin, defaultValues, validator, problemTypes, 'solverParams', true);

Error in integrateRAS (line 32)
[Fmin,FMax] = fluxVariability(modelPop, 100, '', RxnList); %controllare che 100 vada bene

Error in single2IntPopModel (line 87)
    scStruct = integrateRAS(scStruct, 'modelDel');
 
getting this error I am not able to understand please explain.
